### PR TITLE
Revert "Revert "add tags to client metrics" (#590)"

### PR DESCRIPTION
**Pull request by ellisjoe**
This reverts commit d1e472729a8eb71c93a5bfdaf8c761a2c493e799.

### CODE REVIEW

**Comment by ellisjoe** [comment]
opening this to bring back tags. should hold off on merging until we finalize how we're going to serialize tags

**Comment by markelliot** [comment]
Also want to block merging until we have a full release of tritium.

On Oct 27, 2017, at 10:30, Joe Ellis <notifications@github.com<mailto:notifications@github.com>> wrote:


opening this to bring back tags. should hold off on merging until we finalize how we're going to serialize tags

File: okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/metrics/HostMetrics.java
Comment (ellisjoe) [inline okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/metrics/HostMetrics.java]: should be appended to `client.response.`

File: jaxrs-clients/versions.lock
Comment (uschi2000) [inline jaxrs-clients/versions.lock]: what's the point of tritium-api if I need to depend on this whole transitive beast?

File: okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/InstrumentedInterceptor.java
Comment (uschi2000) [inline okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/InstrumentedInterceptor.java:33]: TaggedMetricRegistry should really be an interface and part of the tritium-api so that we can avoid transitive dependencies on all the tritium implementation details. http-remoting already being a shared library, those dependencies make me a little uneasy. in theory, after this merges, tritium can no longer change any of its public APIs.

Comment (uschi2000) [reply okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/InstrumentedInterceptor.java:33]: @schlosna , what would it take to make that change?

Comment (ellisjoe) [reply okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/InstrumentedInterceptor.java:33]: unless we plumb the registry all the way through the remoting apis we're still going to have the issue. explicitly the issue is that we use `TaggedMetricRegistry.getDefault()` to get the global registry at the moment.

Comment (markelliot) [reply okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/InstrumentedInterceptor.java:33]: Can do the serviceloader thing to pull the implementation if we need to.

Comment (ellisjoe) [reply okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/InstrumentedInterceptor.java:33]: ugh

**Comment by uschi2000** [comment]
The only hope for that pattern is slf4j style discovery.

Sent from Email+ secured by MobileIron

-------- Original Message --------

**Comment by uschi2000** [comment]
I know it's pedantic, but these chained dependencies can really lead to a world of pain, so I'd prefer taking the smallest possible dependency.

Sent from Email+ secured by MobileIron

**Comment by ellisjoe** [comment]
In this instance this feels a bit over engineered. The registry is just a wrapper around a map and I don't expect that to change. I'm happy with moving the registry into it's own dependency-less project, but service loading seems like overkill.

**Comment by ellisjoe** [comment]
Also happy to split out an interface, just don't think there's a need to avoid depending on the impl for now.

**Comment by markelliot** [comment]
How do you get the meters and what not without the impl?

**Comment by uschi2000** [comment]
OK, if you're confident that this won't break, go ahead. Can you please add some pretty loud warning signs in tritium that those APIs must not change unless the Java package is incremented?

Sent from Email+ secured by MobileIron

-------- Original Message --------

**Comment by ellisjoe** [comment]
blocked on https://github.com/palantir/tritium/pull/53

**Comment by schlosna** [comment]
[Tritium 0.8.0 released](https://bintray.com/palantir/releases/tritium/0.8.0) with TaggedMetricRegistry interface change and minimized dependencies in `tritium-registry` module

File: okhttp-clients/src/test/java/com/palantir/remoting3/okhttp/InstrumentedInterceptorTest.java
Comment (uschi2000) [inline okhttp-clients/src/test/java/com/palantir/remoting3/okhttp/InstrumentedInterceptorTest.java:70]: nit: check that meter for hostb has count 0

Comment (uschi2000) [reply okhttp-clients/src/test/java/com/palantir/remoting3/okhttp/InstrumentedInterceptorTest.java:70]: actually, I guess you do that below.

**Review by uschi2000 — APPROVED** [review]

File: okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/metrics/HostMetrics.java
Comment (schlosna) [inline okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/metrics/HostMetrics.java:52]: why aren't we putting service name in a tag?

**Comment by uschi2000** [comment]
So the diamond dependency danger turns out to be pretty real.
https://github.com/palantir/tritium/issues/54

**Pull request merged**
